### PR TITLE
UI: EventSource project implementation to enable blocking queries for service and node listings

### PR DESCRIPTION
**Pull request by johncowen**
This PR enables blocking query support for service and node listing pages.

Please note, this is to be merged down onto https://github.com/hashicorp/consul/pull/5070, and lastly down onto `ui-staging` so we can continue this work and integrate the XHR connection management and enable blocking queries for the rest of the UI.

We plan on making enabling blocking queries a 'post-runtime initialisation' user setting. The work for this is here, yet the settings page for this will be re-enabled in another PR before finally merging to `master`.

For the moment to view the changes please manually add a setting via localStorage:

screenshot 2019-01-24 at 18 27 02

### CODE REVIEW

File: ui-v2/app/instance-initializers/event-source.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
+import config from '../config/environment';
```
Comment (johncowen) [inline ui-v2/app/instance-initializers/event-source.js]: Below here we dynamically create the new EventSource returning repositories. Right now we just need to do this with nodes and services - other models will be added in here to enable blocking queries support for those models.

File: ui-v2/app/mixins/with-event-source.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+        if (this[prop] && typeof this[prop].close === 'function') {
+          this[prop].close();
+          // ember doesn't delete on 'resetController' by default
+          delete this[prop];
+        }
+      });
+    }
+    return this._super(...arguments);
+  },
+});
```
Comment (johncowen) [inline ui-v2/app/mixins/with-event-source.js]: The above uses the controller-lifecycle PR (https://github.com/hashicorp/consul/pull/5056) to automatically cleanup any 'closeable' things, in this case EventSources

File: ui-v2/app/services/repository/type/event-source.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+            .catch(function(e) {
+              // setup the aborted connection restarting
+              // this should happen here to avoid cache deletion
+              const status = get(e, 'errors.firstObject.status');
+              if (status === '0') {
+                // Any '0' errors (abort) should possibly try again, depending upon the circumstances
+              }
+              throw e;
+            });
+        };
```
Comment (johncowen) [inline ui-v2/app/services/repository/type/event-source.js]: The above `cb` here is essentially the content of our `Callable/BlockingEventSource`, this is the code that it called every 'tick'. It uses the original ember-dataset resolving Repository that we already had to do the actual requests.

The code around it gives us the opportunity to centrally deal with a post-initialization on/off setting, plus any other settings or global EventSource related errors. Please note, we're planning on removing the `throttle` here and using a rate limiter of some sort.

File: ui-v2/app/services/repository/type/event-source.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    if (cache === null) {
+      cache = createCache({});
+    }
+  },
+  willDestroy: function() {
+    cache = null;
+  },
+  shouldProxy: function(content, method) {
+    return method.indexOf('find') === 0;
+  },
```
Comment (johncowen) [inline ui-v2/app/services/repository/type/event-source.js]: Any `find` methods will be dealt with by this class (which adds a 'layer' over the original Repository class), any other methods will be dealt with by the original itself without the extra 'layer'.

File: ui-v2/app/services/lazy-proxy.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+export default Service.extend({
+  shouldProxy: function(content, method) {
+    return false;
+  },
+  init: function() {
+    this._super(...arguments);
+    const content = get(this, 'content');
+    for (let prop in content) {
+      (prop => {
```
Comment (DingoEatingFuzz) [inline ui-v2/app/services/lazy-proxy.js]: Why wrap this in an iife? Wouldn't removing the iife result in the same thing?

File: ui-v2/tests/acceptance/dc/list-blocking.feature [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  Scenario:
+    And 3 [Model] models
+    And a network latency of 100
+    When I visit the [Page] page for yaml
+    ---
+      dc: dc-1
+    ---
+    Then the url should be /dc-1/[Url]
+    And pause until I see 3 [Model] models
+    And an external edit results in 5 [Model] models
```
Comment (DingoEatingFuzz) [inline ui-v2/tests/acceptance/dc/list-blocking.feature]: This is neat.

File: ui-v2/app/services/lazy-proxy.js (hunk 5 above)
Comment (johncowen) [reply ui-v2/app/services/lazy-proxy.js]: Hmmm yeah , can't remember exactly. I think it's left over from a different version of that that had some async stuff in there somewhere. I vaguely remember putting this in as I hit that `for` async thing where the variable has already 'moved on', but I'm guessing that was in a previous version of this at some point. Will remove, thanks for spotting!

**Pull request merged**